plugins/out_stackdriver/stackdriver.c

```diff
@@ -729,7 +731,41 @@ static int get_stream(msgpack_object_map map)
     return STREAM_UNKNOWN;
 }
 
-static int pack_json_payload(int operation_extracted, int operation_extra_size,
+static void validate_insert_id(msgpack_object * subobj, const msgpack_object * o, 
+                               insert_id_status * status)
+{
+    int i = 0;
+    msgpack_object_kv * p = NULL;
+    *status = INSERTID_NOT_PRESENT;
+
+    if (o == NULL || subobj == NULL) {
+        return;
+    }
+
+    for (i = 0; i < o->via.map.size; i++) {
+        p = &o->via.map.ptr[i];
+        if (p->key.type != MSGPACK_OBJECT_STR) {
+            continue;
+        }
+        if (sizeof(INSERTID_IN_JSON) - 1 == p->key.via.str.size 
+            && strncmp(INSERTID_IN_JSON, p->key.via.str.ptr, p->key.via.str.size) == 0) {
+            if (p->val.type == MSGPACK_OBJECT_STR && p->val.via.str.size > 0) {
+                *subobj = p->val;
```

Comment:
Shouldn't this use copying instead of assigning? Assignment on struct won't deep copy, and msgpack_obj has all kinds of pointers inside [1].

[1] https://github.com/fluent/fluent-bit/blob/f69bf89899cce97fc4a7ee78e653530ca338478c/lib/msgpack-3.2.0/include/msgpack/object.h#L49-L73

Reply:
I meant maybe one of the copy function which does deep copying. E.g. copy from sds library ?

Reply:
In this situation, I think we don't need to deep copy. Keep a pointer is enough.
Besides, in get_msgpack_obj()(https://github.com/fluent/fluent-bit/blob/1c33da012a20a48e06930d36c72bf54ff83179fc/plugins/out_stackdriver/stackdriver.c#L665) above, the author doesn't use deep copy neither. 

WDYT?